### ISSUE TITLE
Chaining reshapes into a matmul with specific shapes causes nonsensical results to appear.

### ISSUE THREAD

**Issue by sgligorijevicTT** [tenstorrent/tt-metal#22012]
### Describe the bug

Calling matmul with both operands being results of reshapes can result in nonsensical or garbage values being returned as a result. In a real model with real inputs, it results in negative values being returned when inputs are all positive, in the attached unit test infinities appear instead. Notably, running just the matmul with identical inputs uploaded from the CPU provides correct result.

### Steps to reproduce the issue

1. Run the following code:
```
import torch

import ttnn


v405 = torch.ones((12,), dtype=torch.float32)
v406 = torch.ones((1, 12, 4, 4), dtype=torch.float32) * 2

def golden_fn(v405, v406):
    v407 = v405.reshape((12, 1, 1))
    v408 = v406.reshape((12, 1, 16))
    return torch.matmul(v407, v408)

golden = golden_fn(v405, v406)

with ttnn.manage_device(0) as dev:
    v405_ttnn = ttnn.from_torch(v405, dtype=ttnn.float32, layout=ttnn.TILE_LAYOUT, device=dev)
    v406_ttnn = ttnn.from_torch(v406, dtype=ttnn.float32, layout=ttnn.TILE_LAYOUT, device=dev)

    v407_ttnn = ttnn.reshape(v405_ttnn, (12, 1, 1))
    v408_ttnn = ttnn.reshape(v406_ttnn, (12, 1, 16))
    ttnn_res = ttnn.matmul(v407_ttnn, v408_ttnn)
    
    res = ttnn.to_torch(ttnn_res)


def calculate_pcc(a,b):
    a = a.flatten()
    b = b.flatten()
    return torch.corrcoef(torch.stack([a, b]))[0, 1].item()


print("PCC:", calculate_pcc(res, golden))

print("Sanity check:")
print(f"Res: {res[0]}")
print(f"golden: {golden[0]}")
```

2. Observe that pcc is nan and that indexing result gives infinities.

### Expected behavior

PCC is 1 and all twos appear as the output of the unit test

### Please complete the following environment information

Ubuntu 22.04, default ird image.
Metal commit: d53a23d3c5e5226e950d623ea1d9e8fba95e9857

**Comment by mrakitaTT** [comment]
Raising the priority because this causes bad PCC in some of our transformer models, and all the workarounds that we can make on our side are pretty ugly (like decomposing such matmul into multiplies).

Tagging @bbradelTT I believe you are the matmul owner? Can you folks please investigate and give us some ETA, because if this will take much time to fix on your side then we might want to introduce a workaround on our side even though it would be ugly :)

**Comment by bbradelTT** [comment]
It's because of the shapes. You need to pad the tensors via the following right before calling ttnn.matmul:
```
    v407_ttnn = ttnn.fill_implicit_tile_padding(v407_ttnn, 0)
    v408_ttnn = ttnn.fill_implicit_tile_padding(v408_ttnn, 0)
```

@cmaryanTT we're going to have to implement this as part of the op, either via calling fill_implicit_tile_padding or doing the equivalent inside the op.

**Comment by bbradelTT** [comment]
@nsorabaTT please create an inside op solution, similar to what you modified in https://github.com/tenstorrent/tt-metal/pull/20008, except the padding would be 0.

It would be good to make a plan to figure out how to extract that code into a common location that can be called from multiple ops, and then which kernels to call the code from.

**Comment by bbradelTT** [comment]
@mrakitaTT the workaround on your end should be relatively simple.

Something like
```
def ttnn_matmul(a, b):
   return ttnn.matmul(ttnn.fill_implicit_tile_padding(a, 0), ttnn.fill_implicit_tile_padding(b, 0))
```
and then call ttnn_matmul.

Having said that, we will look into fixing this.

Which models is this blocking?

**Comment by mrakitaTT** [comment]
Thank you for quick assistance @bbradelTT! We will look into incorporating this workaround.

>Which models is this blocking?

We've seen this happen in attention blocks inside BERT, RoBERTa, ELECTRA and RoFormer JAX models so far, but it's very possible it happens in some other transformer models too.

**Comment by nsorabaTT** [comment]
I have started working on this today, will update here on the progress.

**Comment by nsorabaTT** [comment]
Triaged this to tttnn.reshape not respecting the default pad value of 0.
Created a sub-issue: https://github.com/tenstorrent/tt-metal/issues/22178

In the end, it comes to the contract between matmul and tiled tensors:
1. If the expectation is that tiled places should have zero by default, no change in matmul is needed.
2. If the expectation is that there is no guarantee on the values in the padded locations, then I will take up this issue and add support for it in all matmul kernels.

**Comment by bbradelTT** [comment]
It's the second one. there is no guarantee on the values in the padded locations.

**Comment by nsorabaTT** [comment]
Ok, so the problem can occur with or without the reshape (whenever the padded values are non-zero). Working on a fix in matmul.

In parallel, the reshape issue needs a fix but can be a lower priority.

**Comment by nsorabaTT** [comment]
PR merged: https://github.com/tenstorrent/tt-metal/pull/22958
